### DISCUSSION TITLE
Replace Bentley login with custom Login

### DISCUSSION THREAD

**Discussion by tusharsable** [iTwin/itwinjs-core#2801] [Q&A]
Hi All, 

We want to replace our benley login wih ours. We have referred the given repo [here Calebmshafer](https://github.com/calebmshafer/itwin-viewer-proxy)
 and have rewrote the mytoken file replacing the signin code for custom token url  and a JWT token from our backend which authorises user `public async signIn(requestContext?: ClientRequestContext): Promise<void> {
      requestContext?.enter();

      var config = {
        headers: { 
          'Authorization': `Bearer ${this._JWTToken}`
        }
      };
            //const res = await axios.post(this._tokenUrl,{},config);
            const res = await axios.post("http://localhost:3030/itwinauth",{},config);
       
      if (!res.data)
        throw Error(`Error signing in: ${res.statusText}`);
      this._accessToken = AccessToken.fromJson(res.data);
      console.log("response Token: ",this._accessToken)
    }`

But I am getting error that access token 
![MicrosoftTeams-image (3)](https://user-images.githubusercontent.com/66006244/143988982-76dd5edc-d33b-42e9-8ff3-2df933643ac5.png)

We have two apps on a single account and multiple models .  which I am accessing here. I am not getting why are we getting this error when I have cloned the react viewer repo and used it replacing the signin code and .env for model and context id

**Comment by calebmshafer** [comment]
@tusharsable how are you setting up the server side of your implementation? Updating the [MyTokenServerAuthClient.ts](https://github.com/calebmshafer/itwin-viewer-proxy/blob/main/react-viewer/src/MyTokenServerAuthClient.ts) to call the server you created which is behind our own authentication layer is the correct approach. However, it seems like you may not be returning the correct access token from the token server to the client-side.

**Reply by fors4ken** [reply]
Hey @calebmshafer ,
I had to debug the Bentley library itself (import { AccessToken } from "@bentley/itwin-client") to check if our AccessToken was incorrect. The access token is correct. This error is shown on the client side just when the code reaches the final rendering part (shown below).

```
<div className="viewer-container">
      {isAuthorized && (
        <Viewer
          contextId={process.env.IMJS_CONTEXT_ID}
          iModelId={process.env.IMJS_IMODEL_ID}
          //changeSetId={process.env.IMJS_CHANGESET_ID}
          authConfig={{ oidcClient: MyTokenServerAuthClient.oidcClient }}
        />
      )}
    </div>
```

I believe we are missing something while setting up this auth flow. Here are a few things that we know we definitely did.

**1. Don't forget to add <CLIENT_ID>@apps.imsoidc.bentley.com to your Project too!**
Do you mean Bentley app by project ? Or the iModel itself ? I added this to both the app and the iModel, just to check if had to do anything with it. Still no luck.

2. > how are you setting up the server side of your implementation? 

Its exactly the same as described in your example. We just have our own JWT auth system in place that protects the itwin token endpoint.

**Reply by calebmshafer** [reply]
> Do you mean Bentley app by project ? Or the iModel itself ?

@fors4ken the client id to add to the project is the one used for the iTwin token endpoint. With the iTwin Token endpoint you would have created a "Service" client id and that needs to be added to each Project with the correct RBAC permissions.  The minimum RBAC  permission needed is the iModelHub View permission. Here's an example,

![Capture](https://user-images.githubusercontent.com/31107829/144457403-4dd46610-ed99-4c5b-9b28-deeed6cd00ad.PNG)

**Reply by calebmshafer** [reply]
Also, would you be able to get the x-correlation-id of the failing request? I can check the status of the request on our side with that information.

**Reply by fors4ken** [reply]
@calebmshafer  We had a chat with Glenn few hours and that's when we came to know the hierarchy of registered apps, iModels and Projects. 
The account which we've been using all along was working well when we were directly working with iModels, we had no idea of projects all this time, until we moved to custom authentication. We uploaded 7 models and 7 projects were created with the same name, which were registered by "imodeljs projects". 
The issue is that this account is NOT an administrator so it cannot Manage Users and add the service-app to the project. See below
![project_c](https://user-images.githubusercontent.com/57709627/144459795-a886361d-c720-4243-bc9a-2411c7bef4fb.JPG)
:
There is no Create New Project option available to this account
![project_c1](https://user-images.githubusercontent.com/57709627/144459903-ba4beed4-7803-4425-9180-13221e8a7e14.JPG)
.
The account itself is not an administrator. Can you tell us how we would mark this account as an admin ?

**Reply by calebmshafer** [reply]
@fors4ken how did you initially create the iModels? Did you go through the developer portal and create them with this url, https://developer.bentley.com/my-imodels/?

If so, that project creation workflow is only for testing and demo purposes which is why you won't have full access to manage the project created. To create a new project I'd suggest either directly using our REST APIs for creation, [here](https://developer.bentley.com/apis/projects/) and [here](https://developer.bentley.com/apis/projects/operations/create-project/)~~, or you can go to [itwin.bentley.com](https://itwin.bentley.com/) sign-in with a Bentley account and click "Register a project".~~

Ignore that second part, that doesn't seem to work properly when I just tried it with an unaffiliated user. However, if you go to the link for the [create project API](https://developer.bentley.com/apis/projects/operations/create-project/) and use the "Try it out" button you can easily get started with a project.

**Reply by fors4ken** [reply]
@calebmshafer Thank you. I'll try this and let you know how it goes.
